### DISCUSSION TITLE
Package rules applied in wrong order if using inherited config

### DISCUSSION THREAD

**Discussion by sichapman** [renovatebot/renovate#31182] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab, 38.64.2

### Please tell us more about your question or problem

(I encountered this on a much bigger scale with extending `group:recommended` but have reduced it to just the one here for simplicity)

If I have the following config:

```yaml
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
      "group:flyway"
    ],
    "packageRules": [
      {
        "matchPackageNames": ["**"],
        "matchUpdateTypes": ["minor"],
        "groupName": "all packages"
      }
    ]
}
```
and the following Gradle file (I don't think it's specific to Gradle, seen this issue with Python too):
```groovy
plugins {
    id 'java-library'
}

repositories {
    mavenCentral()
}

dependencies {
    implementation 'org.apache.commons:commons-lang3:3.16.0'
    implementation 'org.flywaydb:flyway-core:10.16.0'
    runtimeOnly 'org.flywaydb:flyway-database-postgresql:10.16.0'
}

java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(17)
    }
}
```

Since all dependencies are one minor version out of date, I would expect the minor version grouping to override the extended flyway version grouping, and one branch+PR to be created for all 3 dependencies. And indeed this is the case if the config is placed in the project repository.

However if the same config is instead placed in a shared inherited config repository, only the commons-lang goes into the minor versions group PR (branch `renovate/all-packages`) - Flyway gets put in it's own (branch `renovate/flyway`). It seems that the packageRule coming in from the `extends` is being applied after the packagesRules in the inherited config, instead of before.

EDIT: Bonus scenarios tested:

`extends` in inherited, `packageRules` in project = 1 PR

`packageRules` in inherited, `extends` in project = 2 PRs

### Logs (if relevant)

<details><summary>Logs</summary>

[Inherited config log](https://gist.github.com/sichapman/6c1ac9643d1c36e82bd5d3c8f9957aa3)

[Project config log](https://gist.github.com/sichapman/4cd159e11de529acff72a102d909de92)

</details>

**Comment by rarkins** [comment]
Please create a minimal reproduction of the simplest thing going wrong and we look at that first

**Reply by sichapman** [reply]
Done, I think: https://github.com/sichapman/31182

**Reply by rarkins** [reply]
Please run or install the bot to complete the reproduction (demonstrating the PRs)

**Reply by sichapman** [reply]
Bot ran, PRs created

**Reply by rarkins** [reply]
Please provide the full debug logs

**Reply by sichapman** [reply]
Log: https://gist.github.com/sichapman/ff84550e17b6a3947bde945b937be732

**Comment by sichapman** [comment]
After digging around in the code I think I've found the problem and a fix:
I think this block of code
https://github.com/renovatebot/renovate/blob/10537afe095b8d1c43dbb0d4b106c0ec91495341/lib/workers/repository/init/merge.ts#L188-L194
needs to be replicated for `returnConfig.packageRules`. I've tried this locally and it does seem to make the rules get resolved into the correct order and only produce one PR.

**Reply by rarkins** [reply]
Can you clarify for my understanding the following?

If the inherit config has extends=A, rules=B, and the repo config has extends=C, rules=D. Then what order are you seeing them resolved in? e.g. A,C,B,D ?

**Reply by sichapman** [reply]
BACD

```
"packageRules": [
           {
             "groupName": "flywayB", // from inherited packagerules
             "matchPackageNames": ["org.flywaydb:*", "org.flywaydb.flyway:*"]
           },
           {
             "groupName": "flywayA", // from inherited extends
             "matchPackageNames": ["org.flywaydb:*", "org.flywaydb.flyway:*"]
           },
           {
             "groupName": "flywayC", // from repo config extends
             "matchPackageNames": ["org.flywaydb:*", "org.flywaydb.flyway:*"]
           },
           {
             "groupName": "flywayD", // from repo config packagerules
             "matchPackageNames": ["org.flywaydb:*", "org.flywaydb.flyway:*"]
           }
         ]
```

**Reply by sichapman** [reply]
With my suggested code addition, it changes to ACBD which sounds like it may be more correct

**Reply by rarkins** [reply]
I think it should be ABCD

**Reply by sichapman** [reply]
Okay in that case scrap my previous suggestion, instead I think `mergeInheritedConfig()` needs an extra call to `resolveConfigPresets`. I don't know the complexities of the code enough to do that properly myself, but I did bodge in a simple call to it and it did change the order to ABCD.

**Comment by rarkins** [comment]
Converted to a bug issue: https://github.com/renovatebot/renovate/issues/31278
